File: algorithms/trees/b-tree/kotlin/BTree.kt
private const val T = 3
private const val MAX_KEYS = 2 * T - 1

private class BTreeNode(var leaf: Boolean = true) {
    val keys = mutableListOf<Int>()
    val children = mutableListOf<BTreeNode>()
}

private fun splitChild(parent: BTreeNode, i: Int) {
    val full = parent.children[i]
    val newNode = BTreeNode(full.leaf)
    val mid = T - 1
    for (j in T until full.keys.size) {
        newNode.keys.add(full.keys[j])
    }
    val median = full.keys[mid]
    if (!full.leaf) {
        for (j in T until full.children.size) {
            newNode.children.add(full.children[j])
        }
        while (full.children.size > T) full.children.removeAt(full.children.size - 1)
    }
    while (full.keys.size > mid) full.keys.removeAt(full.keys.size - 1)
    parent.keys.add(i, median)
    parent.children.add(i + 1, newNode)
}

private fun insertNonFull(node: BTreeNode, key: Int) {
    if (node.leaf) {
        val pos = node.keys.indexOfFirst { it > key }.let { if (it == -1) node.keys.size else it }
        node.keys.add(pos, key)
    } else {
        var i = node.keys.size - 1
        while (i >= 0 && key < node.keys[i]) i--
        i++
        if (node.children[i].keys.size == MAX_KEYS) {
            splitChild(node, i)
            if (key > node.keys[i]) i++
        }
        insertNonFull(node.children[i], key)
    }
}

private fun inorder(node: BTreeNode?, result: MutableList<Int>) {
    if (node == null) return
    for (i in 0 until node.keys.size) {
        if (!node.leaf) inorder(node.children[i], result)
        result.add(node.keys[i])
    }
    if (!node.leaf) inorder(node.children[node.keys.size], result)
}

fun bTree(arr: IntArray): IntArray {
    if (arr.isEmpty()) return intArrayOf()
    var root = BTreeNode(true)
    for (v in arr) {
        if (root.keys.size == MAX_KEYS) {
            val newRoot = BTreeNode(false)
            newRoot.children.add(root)
            splitChild(newRoot, 0)
            root = newRoot
        }
        insertNonFull(root, v)
    }
    val result = mutableListOf<Int>()
    inorder(root, result)
    return result.toIntArray()
}

File: algorithms/trees/b-tree/go/b_tree.go
package btree

const t = 3
const maxKeys = 2*t - 1

type node struct {
	keys     [maxKeys]int
	children [maxKeys + 1]*node
	n        int
	leaf     bool
}

func newNode(leaf bool) *node {
	return &node{leaf: leaf}
}

func splitChild(parent *node, i int) {
	full := parent.children[i]
	nn := newNode(full.leaf)
	nn.n = t - 1
	for j := 0; j < t-1; j++ {
		nn.keys[j] = full.keys[j+t]
	}
	if !full.leaf {
		for j := 0; j < t; j++ {
			nn.children[j] = full.children[j+t]
			full.children[j+t] = nil
		}
	}
	for j := parent.n; j > i; j-- {
		parent.children[j+1] = parent.children[j]
	}
	parent.children[i+1] = nn
	for j := parent.n - 1; j >= i; j-- {
		parent.keys[j+1] = parent.keys[j]
	}
	parent.keys[i] = full.keys[t-1]
	full.n = t - 1
	parent.n++
}

func insertNonFull(nd *node, key int) {
	i := nd.n - 1
	if nd.leaf {
		for i >= 0 && key < nd.keys[i] {
			nd.keys[i+1] = nd.keys[i]
			i--
		}
		nd.keys[i+1] = key
		nd.n++
	} else {
		for i >= 0 && key < nd.keys[i] {
			i--
		}
		i++
		if nd.children[i].n == maxKeys {
			splitChild(nd, i)
			if key > nd.keys[i] {
				i++
			}
		}
		insertNonFull(nd.children[i], key)
	}
}

func inorder(nd *node, result *[]int) {
	if nd == nil {
		return
	}
	for i := 0; i < nd.n; i++ {
		if !nd.leaf {
			inorder(nd.children[i], result)
		}
		*result = append(*result, nd.keys[i])
	}
	if !nd.leaf {
		inorder(nd.children[nd.n], result)
	}
}

// BTree inserts values into a B-Tree and returns sorted inorder traversal.
func BTree(arr []int) []int {
	if len(arr) == 0 {
		return []int{}
	}
	root := newNode(true)
	for _, val := range arr {
		if root.n == maxKeys {
			newRoot := newNode(false)
			newRoot.children[0] = root
			splitChild(newRoot, 0)
			root = newRoot
		}
		insertNonFull(root, val)
	}
	result := []int{}
	inorder(root, &result)
	return result
}
